### ISSUE TITLE
[Regression] TMNT Shredder's Revenge no longer loads on EA builds

### ISSUE THREAD

**Issue by legend800** [yuzu-emu/yuzu#10610]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Affected Build(s)

yuzu Early Access 3635

### Description of Issue

Something in EA currently broke TMNT Shredder's Revenge. It's been broken for a few weeks now. 

Now with Multi-threaded off (only way to get to the opening cinematic), it's permanent black screen on boot with the following error:
Service <Error> core\hle\service\sockets\bsd.cpp:PollImpl:531: File descriptor handle=0 is not allocated

With multi-threaded on, game will boot and you'll get a few splash screens, but it'll freeze/crash at the opening cinematic as it always has. 

Works fine still on Mainline:
<img width="814" alt="2023-06-03_23_39_10-Switch" src="https://github.com/yuzu-emu/yuzu/assets/20983844/60c3a688-95c2-4300-b6ab-e6314b7d7b15">

### Expected Behavior

Game loads on single-threaded mode like it used to.

### Reproduction Steps

Boot game with multi-threaded off.

### Log File

[yuzu_log.txt](https://github.com/yuzu-emu/yuzu/files/11646166/yuzu_log.txt)

### System Configuration

CPU: AMD5900
GPU/Driver: nvidia 3070
RAM: 32gb
OS: Win11

**Comment by Squall-Leonhart** [comment]
It hasn't frozen, its just taking longer to boot.

**Comment by legend800** [comment]
Mainline is not affected. I know what I'm doing here. And I'm not talking about the cinematic. I know you have to skip that. I'm talking about boot/initial splash screens.

Looks like it does eventually load the splash screen if you have give it literally 2-3mins. BUT even once you skip the opening cinematic, the game is significantly slower to load screens (1+ to boot gameplay) and in-game the speed is 70% slower. 

So something in EA really affected single threaded emulation. Everything is perfect in Mainline - loads are snappy with full speed. 

EA:
<img width="814" alt="EA" src="https://github.com/yuzu-emu/yuzu/assets/20983844/6e2708ad-f9b9-4c17-9414-aef2e8a1efa8">
[EA.txt](https://github.com/yuzu-emu/yuzu/files/11646815/EA.txt)

Mainline:
<img width="815" alt="Mainline" src="https://github.com/yuzu-emu/yuzu/assets/20983844/9af92d74-6073-4a3a-a548-44d9eafc6767">
[Mainline.txt](https://github.com/yuzu-emu/yuzu/files/11646814/Mainline.txt)

**Comment by Squall-Leonhart** [comment]
trouble is i've gone so far back now on EA that the changes are definitely in mainline.

Timing fluke is all it could be.

**Comment by legend800** [comment]
All results are consistent. The screenshots above prove the difference. EA is slow (loads, gameplay). Mainline has zero issues (outside of the opening cinematic - codec issue or something).

**Comment by Squall-Leonhart** [comment]
Found the missing PR between mainline and EA, 

https://github.com/yuzu-emu/yuzu/pull/10086

**Comment by liamwhite** [comment]
@Morph1984 see above

**Comment by legend800** [comment]
Nice, the timing of that PR lines up when it seems to break also.

**Comment by Squall-Leonhart** [comment]
3519 is indeed working, so the culprit seems clear

**Comment by Turdpeterson** [comment]
seems like the issue has made it way into mainline build as this game no longer will boot

**Comment by legend800** [comment]
Yes, it was merged in without a fix for this about 2 weeks ago. No note as to what's the next step to unbreak compatibility with this popular game.

**Comment by liamwhite** [comment]
Optimally, we fix it in multicore. I think there is a race condition in the game that is breaking it, rather than a bug in yuzu, but it's hard to prove the existence of this, and even harder to come up with a fix for those.

**Comment by legend800** [comment]
TBF, there is something screwy with that intro cinematic. 

On Ryu, the only way to bypass it is with Vulkan (or OGL w/gfx multi-threading off) + Vsync off. Even with that, the cinematic is choppy. 

Either of those not set like that, it just hangs too. Setting the equivalent on yuzu doesn't help - only single core cpu on older builds gets it past that video.

**Comment by liamwhite** [comment]
Here is a mod for 1.0.6 that allows it to boot to the menu in yuzu with multicore on. I imagine it would also fix other modes on Ryujinx, but I didn't test that.

[TMNT_Shredders_Revenge_1.0.6.boot.zip](https://github.com/yuzu-emu/yuzu/files/12178660/TMNT_Shredders_Revenge_1.0.6.boot.zip)

There is a race condition that happens on startup, where the game spawns a new thread for the video player. Because there is no locking and none of the data structures are threadsafe, the game ends up concurrently trying to start a render from the video player thread and complete a render from the main thread, which throws an exception and crashes the process. This happens no matter how much the latency between the thread spawn and thread start is increased.

On console, this doesn't crash because the timing works out that the video playback rendering will always happen in between updates on the main thread. Skewing the timing even slightly causes it to crash there as well, so one would hope Dotemu will eventually fix this bug.

Whether or not the game will use a separate thread to spawn the video player is controlled by a single flag in the executable--probably because they were having trouble with this in development--and this mod simply prevents that flag from getting set. This makes sure it doesn't spawn the thread and runs the video player on the main thread. (The video playback is very broken, however, which is another bug in the game.)

**Comment by legend800** [comment]
Works perfect (well minus the broken cinematic)! Great job on the investigation and thanks for sharing the fix. This does explain why yuzu/ryu have always had issues with getting past that cinematic. 

Hopefully Dotemu fixes this on 1.0.7, else we may need an updated patch :). Seems doubtful though since there's been several updates and if it doesn't affect console users, why would they bother. We'll see. 

The cinematic does play relatively ok on Ryu, especially after a few sec. What's the bug with the video playback itself?

**Comment by liamwhite** [comment]
I don't really know why it's breaking the way it does--the game checks to see if the time spent on the CPU was within expected boundaries, and if it wasn't, it draws the color black to the screen instead of drawing a frame. It's still decoding the video, it's just not presenting the frame because it doesn't think it's supposed to present it. I have never seen this type of behavior before.

**Comment by ankur20us** [comment]
how to install it in yuzu 

> Here is a mod for 1.0.6 that allows it to boot to the menu in yuzu with multicore on. I imagine it would also fix other modes on Ryujinx, but I didn't test that.
> 
> [TMNT_Shredders_Revenge_1.0.6.boot.zip](https://github.com/yuzu-emu/yuzu/files/12178660/TMNT_Shredders_Revenge_1.0.6.boot.zip)
> 
> There is a race condition that happens on startup, where the game spawns a new thread for the video player. Because there is no locking and none of the data structures are threadsafe, the game ends up concurrently trying to start a render from the video player thread and complete a render from the main thread, which throws an exception and crashes the process. This happens no matter how much the latency between the thread spawn and thread start is increased.
> 
> On console, this doesn't crash because the timing works out that the video playback rendering will always happen in between updates on the main thread. Skewing the timing even slightly causes it to crash there as well, so one would hope Dotemu will eventually fix this bug.
> 
> Whether or not the game will use a separate thread to spawn the video player is controlled by a single flag in the executable--probably because they were having trouble with this in development--and this mod simply prevents that flag from getting set. This makes sure it doesn't spawn the thread and runs the video player on the main thread. (The video playback is very broken, however, which is another bug in the game.)

How to install this zip file in yuzu or Ryujinx given the condition that we are using both the emulators in Retrobat? Please guide. i really want to play this game but game is not running :(

**Comment by natebizzle** [comment]
> Works perfect (well minus the broken cinematic)! Great job on the investigation and thanks for sharing the fix. This does explain why yuzu/ryu have always had issues with getting past that cinematic.
> 
> Hopefully Dotemu fixes this on 1.0.7, else we may need an updated patch :). Seems doubtful though since there's been several updates and if it doesn't affect console users, why would they bother. We'll see.
> 
> The cinematic does play relatively ok on Ryu, especially after a few sec. What's the bug with the video playback itself?

How are you even getting this to load past the intro credits? I have the 1.0.6 update and the abovementioned mod both installed but the game just crashes at a blank white cinematic screen now and yuzu quits. Prior to installing the mod, the game wouldn't even load on multi-core, but single core would get to a blank black cinematic screen and freeze then I'd have to force quit.

**Comment by Squall-Leonhart** [comment]
> How are you even getting this to load past the intro credits? I have the 1.0.6 update and the abovementioned mod both installed but the game just crashes at a blank white cinematic screen now and yuzu quits. Prior to installing the mod, the game wouldn't even load on multi-core, but single core would get to a blank black cinematic screen and freeze then I'd have to force quit.

exit yuzu and try again until it works basically.

**Comment by legend800** [comment]
> Here is a mod for 1.0.6 that allows it to boot to the menu in yuzu with multicore on. I imagine it would also fix other modes on Ryujinx, but I didn't test that.
> 
> [TMNT_Shredders_Revenge_1.0.6.boot.zip](https://github.com/yuzu-emu/yuzu/files/12178660/TMNT_Shredders_Revenge_1.0.6.boot.zip)

@liamwhite Sorry for the bother, but 1.0.7 just came out (for Dimension Shellshock DLC) but still same boot issues. I tried just renaming your file to 1.0.7 but the mem/id is diff. Any chance you can update the flag disable mod to support the latest version?

**Comment by legend800** [comment]
Also, I just found that the opening cinematic renders **_perfectly_** with OGL, as opposed to only black in Vulkan.

<img width="812" alt="2023-08-31 18_46_07-" src="https://github.com/yuzu-emu/yuzu/assets/20983844/de03a1e7-b964-4bdf-842b-342e8b556f47">

**Comment by liamwhite** [comment]
rebased patch [TMNT_Shredders_Revenge_1.0.7.boot.zip](https://github.com/yuzu-emu/yuzu/files/12494211/TMNT_Shredders_Revenge_1.0.7.boot.zip)

**Comment by legend800** [comment]
Works great! 

<img width="813" alt="2023-08-31 21_23_55-yuzu Early Access 3851 _ Teenage Mutant Ninja Turtles_ Shredder's Revenge (64-bi" src="https://github.com/yuzu-emu/yuzu/assets/20983844/e26f9d2f-0fb4-4fb1-a231-95ad07039961">

Though with 1.0.7, the game won't start (stays on yuzu "launching" or first black screen forever) like 75% of boot attempts. That's new. :/

**Comment by Squall-Leonhart** [comment]
>  Perm black screen after char select no matter what I try (ogl/vk, single/multi core, etc.). Just spams this stuff forever.

seems this isn't limited to 1.0.7, but i do not have the log spam.

**Comment by legend800** [comment]
That issue is limited to 1.0.7. Never been a problem on 1.0.0-1.0.6. 
https://github.com/yuzu-emu/yuzu/issues/11440

**Comment by Squall-Leonhart** [comment]
Nope, 1.0.6 with the patch is black screening after character select.

**Comment by legend800** [comment]
Interesting. Tested again and yep, it's the boot fix @liamwhite . Looks like that fixes the opening cinematic **_but_** breaks loading in-game map unfortunately. This is the same for either 1.06 or 1.07 TU/boot fix. Any chance you can take a look?

So it looks like the ONLY way to play this game on Yuzu is to use the June 21st build, before the cpu change was made noted above. With that build (and no "boot fix" mod), you can enter and play the game fine (minus choppy cinematic), including the latest version.

<img width="814" alt="2023-10-02 23_22_54-yuzu 1475 _ Teenage Mutant Ninja Turtles_ Shredder's Revenge (64-bit) _ 1 0 7 _ " src="https://github.com/yuzu-emu/yuzu/assets/20983844/fa255d16-9cd3-4507-ad66-8c751b3bf82d">

Settings:
- Single Core(boot)
- VK(flickering on old builds like June 21 with OGL)

**Comment by Squall-Leonhart** [comment]
I suspect the memory patch needs to be made conditional and return the original value once video is shut down.

**Comment by musgrove-ernest** [comment]
so i am having a problem with both TMNT games not working on the steamdeck.  please advise what files you need, and i will try to find them and post them.  and explain not working i get and error on both games and i have to do a hard restart of the deck and then switch from normal mode to desktop mode or desktop mode to normal mode, just to get yuzu to start working again

**Comment by legend800** [comment]
@musgrove-ernest That has nothing to do with this issue. Go get support on Discord or create a new issue.

**Comment by edvin76** [comment]
> rebased patch [TMNT_Shredders_Revenge_1.0.7.boot.zip](https://github.com/yuzu-emu/yuzu/files/12494211/TMNT_Shredders_Revenge_1.0.7.boot.zip)

hey it still doesnt work, it goes to the main menu but starting the game crashes

as previousy said:
"I suspect the memory patch needs to be made conditional and return the original value once video is shut down."

**Comment by Challebjoern** [comment]
Does anyone know why I´m not even getting the game to show up in Yuzu?

**Comment by edvin76** [comment]
that's normally prod key issue

**Comment by fartboygh** [comment]
Is anyone working on this?

**Comment by mania9860** [comment]
It runs on yuzu android!!! #
![스크린샷 2023-11-25 222233](https://github.com/yuzu-emu/yuzu/assets/151451384/8d7a9ee2-280a-42c7-af2c-5d9862878753)

**Comment by kathyrollo** [comment]
It appears this issue now also occuring at with the mainline build 1674. It is stuck on this screen:

![Screenshot 2024-01-10 154158](https://github.com/yuzu-emu/yuzu/assets/28589393/8e528f34-b504-4fd7-aaa0-81a363092db0)

DLC + Update 1.0.8 enabled. FW 16.0.3

**Comment by legend800** [comment]
Last Mainline build that worked for this game was 1475. The next build had cpu changes from EA that broke it. Needs the below config and everything works for all versions. Only way to play the game properly on Yuzu. 

Config:
- Single core (boot/intro vid fix)
- Native res (fix small video)
- VK (flickering)
- Gaussian filter (preference)

**Comment by kathyrollo** [comment]
Thanks @legend800 . Will try your suggested config when I get back. What does VK mean?

**Comment by fartboygh** [comment]
> Last Mainline build that worked for this game was 1475. The next build had cpu changes from EA that broke it. Needs the below config and everything works for all versions. Only way to play the game properly on Yuzu.
> 
> Config:
> 
>     * Single core (boot/intro vid fix)
> 
>     * Native res (fix small video)
> 
>     * VK (flickering)
> 
>     * Gaussian filter (preference)

Can't find VK (flickering)

**Comment by legend800** [comment]
Vulkan

**Comment by fartboygh** [comment]
Doesn't work.

**Comment by kathyrollo** [comment]
> Last Mainline build that worked for this game was 1475. The next build had cpu changes from EA that broke it. Needs the below config and everything works for all versions. Only way to play the game properly on Yuzu.
> 
> Config:
> 
> * Single core (boot/intro vid fix)
> * Native res (fix small video)
> * VK (flickering)
> * Gaussian filter (preference)

Thanks, I got it working:

![Screenshot 2024-01-17 173056](https://github.com/yuzu-emu/yuzu/assets/28589393/af1cb8b9-dbae-4881-b7f4-255849cf8d11)

![Screenshot 2024-01-17 172955](https://github.com/yuzu-emu/yuzu/assets/28589393/3bd3e5b6-e570-48aa-8d77-2156f2d12b3e)

CPU: Ryzen 7 5800H (8C/16T)
GPU / Driver: Radeon Graphics (integrated) / Ver. 23.11.1 
RAM / VRAM: 32GB / 8GB (allocated)

Build: Yuzu ML 1680
FW: 16.0.3
Add-Ons: DLC (enabled), Update 1.0.8 (enabled)

The intro video is all black with stuttery audio, so I just hit (+) button to skip it.

**Comment by SamdroidVa** [comment]
Playing this game without Multicore is disappointing. Framerate takes a nasty hit.  I suggest using Ryujinx, the game runs flawless. Until then I'll be praying for a fix.

**Comment by legend800** [comment]
UPDATE:
The game devs finally did something to the quirky opening cinematic on the latest 1.0.8 TU, and it no longer hangs there on Yuzu (note only renders correctly on OGL). 

BUT, you still need single core to boot the game. And unfortunately, it's slower on latest EA build than from last year's build 1475. With the former it's around 85-90% speed, and 100% on 1475. So use the old build for now. 

Still, good progress. If we can just get this game to boot on multi-core, we'll be all set!

<img width="814" alt="2024-02-10 00_50_32-yuzu Early Access 4128 _ Teenage Mutant Ninja Turtles_ Shredder's Revenge (64-bi" src="https://github.com/yuzu-emu/yuzu/assets/20983844/c465a7d9-a2ef-4b05-9365-a68e2ad9c0bf">

**Comment by kathyrollo** [comment]
Hi @legend800 , OGL performs badly on my Ryzen 7 5800H mini-pc. Still doesn't work for Vulkan you reckon?
